drivers/sensor/current_amp/current_amp.c

```diff
@@ -59,16 +59,23 @@ static int get(const struct device *dev, enum sensor_channel chan, struct sensor
 		return ret;
 	}
 
-	i_ma = raw_val;
+#ifdef CONFIG_CURRENT_AMP_HIGH_RANGE
+	int32_t i_ma = raw_val;
+
 	current_sense_amplifier_scale_dt(config, &i_ma);
+	i_ua = 1000 * i_ma;
+#else
+	/* Limited to +-32.7V */
+	__ASSERT(raw_val <= INT16_MAX, "Voltage outside of range");
+	__ASSERT(raw_val >= INT16_MIN, "Voltage outside of range");
 
-	LOG_DBG("%d/%d, %dmV, current:%dmA", data->raw,
-		(1 << data->sequence.resolution) - 1, raw_val, i_ma);
+	i_ua = current_sense_amplifier_scale_ua_dt(config, raw_val);
+#endif
 
-	val->val1 = i_ma / 1000;
-	val->val2 = (i_ma % 1000) * 1000;
+	LOG_DBG("%d/%d, %dmV, current:%dmA", data->raw, (1 << data->sequence.resolution) - 1,
+		raw_val, i_ua / 1000);
 
-	return 0;
+	return sensor_value_from_micro(val, i_ua);
```

Comment:
This seems like you can do this at runtime, the equation is the same just scaled by 1,000.

Reply:
In theory yes, but that 1000x scaling means the data types need to change in order to not overflow.
I don't believe this can be done properly without both code paths existing.

Reply:
How much are we saving here? Why can't we just always use `int64_t`?

Reply:
The problem is not the size of the variable here, its the size of the internal variable in `current_sense_amplifier_scale_ua_dt`. If the input range there changes from `int16_t` to `int32_t`, the maximum value of `scaled` exceeds `INT64_MAX`. This could be mitigated if `max_gain` went from the `uint16_t` to a `uint8_t`, but thats changing the allowed range on the devicetree property.

Reply:
I think we might have misunderstood eachother. What's stopping us from doing this...

```
if (raw_val >= INT16_MIN && raw_val <= INT16_MAX) {
  i_ua = current_sense_amplifier_scale_ua_dt(config, raw_val);
} else {
  int i_ma = raw_val;

  current_Sense_amplifier_scale_dt(config, &i_ma);
  i_ua = 1000 * i_ma;
}
```

Reply:
Only the suspicion that reading a raw voltage of over 32V in a current amplifier driver is so far out of range of the expected values that the code will in practice never run. It was preserved as a compile time option in case that assumption is wrong.

Reply:
The ~32V limit is at the ADC input (V_shunt) right?

Reply:
Yes, the ADC input voltage, as determined from the raw units by `adc_raw_to_millivolts_dt`.